Libplanet/Store/DefaultStore.cs

```diff
@@ -246,174 +271,237 @@ public override IEnumerable<TxId> IterateStagedTransactionIds()
-                }
-
-                using (var stream = new MemoryStream())
-                {
-                    DownloadFile(file, stream);
+                return (Transaction<T>)cachedTx;
+            }
 
-                    bytes = stream.ToArray();
-                    if (bytes.Length != file.Length || bytes.Length < 1)
-                    {
-                        _logger.Warning(
-                            "The data file for the transaction {TxId} seems corrupted; " +
-                            "it will be treated nonexistent and removed at all.",
-                            txid
-                        );
-                        _txLock.EnterWriteLock();
-                        try
-                        {
-                            _db.FileStorage.Delete(fileId);
-                        }
-                        finally
-                        {
-                            _txLock.ExitWriteLock();
-                        }
+            UPath path = TxPath(txid);
+            if (!_txs.FileExists(path))
+            {
+                return null;
+            }
 
-                        return null;
-                    }
-                }
+            byte[] bytes;
+            try
+            {
+                bytes = _txs.ReadAllBytes(path);
             }
-            finally
+            catch (FileNotFoundException)
             {
-                _txLock.ExitUpgradeableReadLock();
+                return null;
             }
 
-            return Transaction<T>.FromBencodex(bytes);
+            Transaction<T> tx = Transaction<T>.FromBencodex(bytes);
+            _txCache.AddOrUpdate(txid, tx);
+            return tx;
         }
 
         /// <inheritdoc/>
         public override void PutTransaction<T>(Transaction<T> tx)
         {
-            string fileId = TxFileId(tx.Id);
-            string filename = tx.Id.ToHex();
-            byte[] txBytes = tx.ToBencodex(true);
-            _txLock.EnterUpgradeableReadLock();
-            try
+            if (_txCache.ContainsKey(tx.Id))
             {
-                LiteFileInfo file = _db.FileStorage.FindById(fileId);
-                if (file is LiteFileInfo)
-                {
-                    // No-op if already exists.
-                    return;
-                }
+                return;
+            }
+
+            byte[] txBytes = tx.ToBencodex(true);
+            UPath path = TxPath(tx.Id);
+            UPath dirPath = path.GetDirectory();
+            CreateDirectoryRecursively(_txs, dirPath);
 
-                _txLock.EnterWriteLock();
+            if (_root is MemoryFileSystem)
+            {
+                _txs.WriteAllBytes(path, txBytes);
+            }
+            else
+            {
+                // For atomicity, writes bytes into an intermediate temp file,
+                // and then renames it to the final destination.
+                UPath tmpPath = dirPath / $".{Guid.NewGuid():N}.tmp";
```

Comment:
It seems to possible we abstract codes with creating a method that creates file atomically.

Reply:
Fixed: 43c1ef3e82f7b50ae1f06bbc684c3a746bf888f5.